lib/ansible/modules/cloud/amazon/ec2_scaling_policy.py

```diff
@@ -31,155 +41,353 @@
-            connection.delete_policy(sp_name, asg_name)
+            connection.delete_policy(AutoScalingGroupName=policy['ScalingPolicies'][0]['AutoScalingGroupName'],
+                                     PolicyName=policy_name)
             module.exit_json(changed=True)
-        except BotoServerError as e:
-            module.exit_json(changed=False, msg=str(e))
+        except botocore.exceptions.ClientError as e:
+            module.fail_json(msg="Failed to delete autoscaling policy",
+                             exception=traceback.format_exc(), **camel_dict_to_snake_dict(e.response))
     else:
         module.exit_json(changed=False)
 
 
 def main():
+    step_adjustment_spec = dict(
+        lower_bound=dict(type='int'),
+        upper_bound=dict(type='int'),
+        scaling_adjustment=dict(type='int', required=True))
+
     argument_spec = ec2_argument_spec()
     argument_spec.update(
         dict(
-            name=dict(required=True, type='str'),
-            adjustment_type=dict(type='str', choices=['ChangeInCapacity', 'ExactCapacity', 'PercentChangeInCapacity']),
-            asg_name=dict(required=True, type='str'),
+            name=dict(required=True),
+            adjustment_type=dict(choices=['ChangeInCapacity', 'ExactCapacity', 'PercentChangeInCapacity']),
+            asg_name=dict(),
             scaling_adjustment=dict(type='int'),
             min_adjustment_step=dict(type='int'),
             cooldown=dict(type='int'),
             state=dict(default='present', choices=['present', 'absent']),
+            metric_aggregation=dict(default='Average', choices=['Average', 'Maximum', 'Minimum']),
+            policy_type=dict(default='SimpleScaling', choices=['SimpleScaling', 'StepScaling']),
+            step_adjustments=dict(type='list', options=step_adjustment_spec),
+            estimated_instance_warmup=dict(type='int')
         )
     )
 
-    module = AnsibleModule(argument_spec=argument_spec)
-
-    if not HAS_BOTO:
-        module.fail_json(msg='boto required for this module')
+    module = AnsibleModule(argument_spec=argument_spec,
+                           required_if=[['state', 'present', ['asg_name', 'adjustment_type']]])
 
-    region, ec2_url, aws_connect_params = get_aws_connection_info(module)
+    if not HAS_BOTO3:
+        module.fail_json(msg='boto3 and botocore are required for this module')
 
-    state = module.params.get('state')
+    region, ec2_url, aws_connect_params = get_aws_connection_info(module, boto3=True)
 
     try:
-        connection = connect_to_aws(boto.ec2.autoscale, region, **aws_connect_params)
-    except (boto.exception.NoAuthHandlerFound, AnsibleAWSError) as e:
-        module.fail_json(msg=str(e))
+        connection = boto3_conn(module, conn_type='client', resource='autoscaling',
+                                region=region, endpoint=ec2_url, **aws_connect_params)
+    except (botocore.exceptions.NoCredentialsError, botocore.exceptions.ProfileNotFound) as e:
```

Comment:
The connection doesn't need exception handling in the module. You could also update this to use AnsibleAWSModule and connection = module.client('autoscaling') if you wanted.

Reply:
Yeah, AnsibleAWSModule didn't exist when I made this PR :)